src/Storages/StorageView.cpp

```diff
@@ -87,60 +84,53 @@ Pipes StorageView::read(
     return pipes;
 }
 
-ASTPtr StorageView::getRuntimeViewQuery(const ASTSelectQuery & outer_query, const Context & context)
-{
-    auto temp_outer_query = outer_query.clone();
-    auto * new_outer_select = temp_outer_query->as<ASTSelectQuery>();
-    return getRuntimeViewQuery(new_outer_select, context, false);
-}
-
-
-static void replaceTableNameWithSubquery(ASTSelectQuery * select_query, ASTPtr & subquery)
+static ASTTableExpression * getFirstTableExpression(ASTSelectQuery & select_query)
 {
-    auto * select_element = select_query->tables()->children[0]->as<ASTTablesInSelectQueryElement>();
+    auto * select_element = select_query.tables()->children[0]->as<ASTTablesInSelectQueryElement>();
 
     if (!select_element->table_expression)
         throw Exception("Logical error: incorrect table expression", ErrorCodes::LOGICAL_ERROR);
 
-    auto * table_expression = select_element->table_expression->as<ASTTableExpression>();
+    return select_element->table_expression->as<ASTTableExpression>();
+}
+
+ASTPtr StorageView::replaceWithSubquery(ASTSelectQuery & select_query, ASTPtr & view_name) const
+{
+    ASTPtr subquery = inner_query->clone();
+    ASTTableExpression * table_expression = getFirstTableExpression(select_query);
 
     if (!table_expression->database_and_table_name)
         throw Exception("Logical error: incorrect table expression", ErrorCodes::LOGICAL_ERROR);
 
-    const auto alias = table_expression->database_and_table_name->tryGetAlias();
+    DatabaseAndTableWithAlias db_table(table_expression->database_and_table_name);
+    String alias = db_table.alias.empty() ? db_table.table : db_table.alias;
+
+    view_name = table_expression->database_and_table_name;
     table_expression->database_and_table_name = {};
     table_expression->subquery = std::make_shared<ASTSubquery>();
     table_expression->subquery->children.push_back(subquery);
-    table_expression->children.push_back(table_expression->subquery);
-    if (!alias.empty())
-        table_expression->subquery->setAlias(alias);
-}
+    table_expression->subquery->setAlias(alias);
 
+    for (auto & child : table_expression->children)
+        if (child.get() == view_name.get())
+            child = subquery;
+    return subquery;
+}
 
-ASTPtr StorageView::getRuntimeViewQuery(ASTSelectQuery * outer_query, const Context & context, bool normalize)
```

Comment:
Remove additional table columns resolution here. Also remove PredicateExpressionsOptimizer call. There would be only one in SyntaxAnalyzer. Optimize VIEW subselect in InterpreterSelectQuery instead: full `analyze()` call, not only push down.

Negative side effect - ANALYZE for VIEW shows not optimized view query.